# [Assignment 3](https://ovgu-ailab.github.io/idl2023/assignment3.html)

**_Colaborative Work from Adrian Bremer and Philipp Reinig_**

## Setup

In [15]:
import tensorflow as tf
import keras
from keras import layers
import numpy as np


## CNN for MNIST
### Prepare data

In [16]:
(train_imgs, train_lbls), (test_imgs, test_lbls) = keras.datasets.mnist.load_data()

train_data = tf.data.Dataset.from_tensor_slices(
    (train_imgs.reshape((-1,28,28,1)).astype(np.float32) / 255, train_lbls.astype(np.int32))).shuffle(buffer_size=60000).batch(128).repeat()

test_data = tf.data.Dataset.from_tensor_slices(
    (test_imgs.reshape((-1,28,28,1)).astype(np.float32) / 255, test_lbls.astype(np.int32))).batch(128)


### Build CNN model with Keras

In [31]:
#keras.Input
#layers.Conv2D
#layers.GlobalMaxPool[ing]2D
#layers.MaxPool[ing]2D
#layers.Flatten
#layers.Dense
#keras.Model

model = keras.Sequential([
    layers.Conv2D(filters=16, kernel_size=5, padding="same", activation="relu"),
    layers.MaxPooling2D(),
    layers.Conv2D(filters=32, kernel_size=5, padding="valid", activation="relu"),
    layers.MaxPooling2D(),
    layers.Flatten(),
    layers.Dense(units=128, activation="relu"),
    layers.Dense(units=10)
], name="simple_conv")

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])


In [27]:
model.fit(x=train_data, epochs=10, steps_per_epoch=50)

Epoch 1/10
50/50 [==============================] - 2s 29ms/step - loss: 0.9006 - accuracy: 0.7511
Epoch 2/10
50/50 [==============================] - 1s 22ms/step - loss: 0.2849 - accuracy: 0.9181
Epoch 3/10
50/50 [==============================] - 1s 20ms/step - loss: 0.1960 - accuracy: 0.9423
Epoch 4/10
50/50 [==============================] - 1s 21ms/step - loss: 0.1615 - accuracy: 0.9534
Epoch 5/10
50/50 [==============================] - 1s 21ms/step - loss: 0.1194 - accuracy: 0.9623
Epoch 6/10
50/50 [==============================] - 1s 21ms/step - loss: 0.1137 - accuracy: 0.9630
Epoch 7/10
50/50 [==============================] - 1s 21ms/step - loss: 0.0856 - accuracy: 0.9745
Epoch 8/10
50/50 [==============================] - 1s 21ms/step - loss: 0.0864 - accuracy: 0.9750
Epoch 9/10
50/50 [==============================] - 1s 21ms/step - loss: 0.0754 - accuracy: 0.9803
Epoch 10/10
50/50 [==============================] - 1s 22ms/step - loss: 0.0730 - accuracy: 0.9785


In [28]:
model.summary()

Model: "simple_conv"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_12 (Conv2D)          (None, 28, 28, 16)        416       
                                                                 
 max_pooling2d_12 (MaxPooli  (None, 14, 14, 16)        0         
 ng2D)                                                           
                                                                 
 conv2d_13 (Conv2D)          (None, 10, 10, 32)        12832     
                                                                 
 max_pooling2d_13 (MaxPooli  (None, 5, 5, 32)          0         
 ng2D)                                                           
                                                                 
 flatten_6 (Flatten)         (None, 800)               0         
                                                                 
 dense_12 (Dense)            (None, 128)               

In [29]:
model.evaluate(test_data)

79/79 [==============================] - 1s 7ms/step - loss: 0.0519 - accuracy: 0.9836


[0.051912419497966766, 0.9836000204086304]

In [30]:
model.metrics_names

['loss', 'accuracy']

## CNN on fashion mnist

### Reusing old model / only changing the datasets

In [32]:
(train_imgs, train_lbls), (test_imgs, test_lbls) = keras.datasets.fashion_mnist.load_data()

train_data = tf.data.Dataset.from_tensor_slices(
    (train_imgs.reshape((-1,28,28,1)).astype(np.float32) / 255, train_lbls.astype(np.int32))).shuffle(buffer_size=60000).batch(128).repeat()

test_data = tf.data.Dataset.from_tensor_slices(
    (test_imgs.reshape((-1,28,28,1)).astype(np.float32) / 255, test_lbls.astype(np.int32))).batch(128)


4422102/4422102 [==============================] - 5s 1us/step


In [33]:
# run the cell where the previous model is defined and compiled
# then run this cell to train on fashion mnist
model.fit(x=train_data, epochs=10, steps_per_epoch=50)

Epoch 1/10
50/50 [==============================] - 2s 21ms/step - loss: 1.1508 - accuracy: 0.5841
Epoch 2/10
50/50 [==============================] - 1s 21ms/step - loss: 0.7096 - accuracy: 0.7347
Epoch 3/10
50/50 [==============================] - 1s 21ms/step - loss: 0.6171 - accuracy: 0.7750
Epoch 4/10
50/50 [==============================] - 1s 21ms/step - loss: 0.5352 - accuracy: 0.8069
Epoch 5/10
50/50 [==============================] - 1s 22ms/step - loss: 0.4949 - accuracy: 0.8211
Epoch 6/10
50/50 [==============================] - 1s 22ms/step - loss: 0.4639 - accuracy: 0.8328
Epoch 7/10
50/50 [==============================] - 1s 22ms/step - loss: 0.4514 - accuracy: 0.8422
Epoch 8/10
50/50 [==============================] - 1s 21ms/step - loss: 0.4552 - accuracy: 0.8308
Epoch 9/10
50/50 [==============================] - 1s 21ms/step - loss: 0.4191 - accuracy: 0.8500
Epoch 10/10
50/50 [==============================] - 1s 22ms/step - loss: 0.4133 - accuracy: 0.8527


In [34]:
model.evaluate(test_data)

79/79 [==============================] - 1s 8ms/step - loss: 0.4295 - accuracy: 0.8469


[0.42950576543807983, 0.8468999862670898]

**Conclusions**:
The same model with only 2 convolutional layers and one dense layer performs quite well on fashion mnist. It reaches an accuracy of around 85%.